flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/rules/common/PushFilterIntoTableSourceScanRuleBase.scala

```diff
@@ -36,36 +43,34 @@ trait PushFilterIntoTableSourceScanRuleBase {
       filterableSource: FilterableTableSource[_],
       description: String): Unit = {
 
-    if (filterableSource.isFilterPushedDown) {
-      // The rule can get triggered again due to the transformed "scan => filter"
-      // sequence created by the earlier execution of this rule when we could not
-      // push all the conditions into the scan
-      return
-    }
+    Preconditions.checkArgument(!filterableSource.isFilterPushedDown)
 
     val program = calc.getProgram
+    val functionCatalog = FunctionCatalog.withBuiltIns
     val (predicates, unconvertedRexNodes) =
       RexProgramExtractor.extractConjunctiveConditions(
         program,
         call.builder().getRexBuilder,
-        tableSourceTable.tableEnv.getFunctionCatalog)
+        functionCatalog)
     if (predicates.isEmpty) {
       // no condition can be translated to expression
       return
     }
 
-    val (newTableSource, remainingPredicates) = filterableSource.applyPredicate(predicates)
-    // trying to apply filter push down, set the flag to true no matter whether
-    // we actually push any filters down.
-    newTableSource.setFilterPushedDown(true)
+    val remainingPredicates = new util.LinkedList[Expression]()
+    predicates.foreach(e => remainingPredicates.add(e))
+
+    val newTableSource = filterableSource.applyPredicate(remainingPredicates)
```

Comment:
I think we don't have to restrict this. If the user for some reason indeed want to change the predicates which returns back and  executed by framework, we should allow them to do so.